### ISSUE TITLE
Failed to parse model from solver.

### ISSUE THREAD

**Issue by gyggg** [hopv/hoice#45]
For this input:

```
(set-logic HORN)

(declare-fun str_invariant
             (Int
              Int
              Bool
              Bool
              Bool
              Bool
              Bool
              Int
              Int
              Int
              Int
              Bool
              Bool
              Bool
              Bool
              Bool)
             Bool)
(assert (forall ((A Bool)
         (B Bool)
         (C Bool)
         (D Bool)
         (E Bool)
         (F Int)
         (G Int)
         (H Int)
         (I Int)
         (J Bool)
         (K Bool)
         (L Bool)
         (M Bool)
         (N Bool)
         (O Int)
         (P Int))
  (=> (and (str_invariant P O N M L K J I H G F E D C B A) (not (= K true)))
         false)
    ))
(assert (forall ((A Bool)
         (B Bool)
         (C Bool)
         (D Bool)
         (E Bool)
         (F Int)
         (G Int)
         (H Int)
         (I Int)
         (J Bool)
         (K Bool)
         (L Bool)
         (M Bool)
         (N Bool)
         (O Int)
         (P Int))
  (let ((a!1 (= E (and (>= I 1) (>= H 1) (=> (>= G I) M) (=> (>= F H) L)))))
       (=> (and (= J true)
                (= C true)
                (= B true)
                (= A true)
                (= K true)
                (= I P)
                (= H O)
                (= G 0)
                (= F 0)
                a!1
                (= D E))
           (str_invariant P O N M L K J I H G F E D C B A)))
    ))
(assert (forall ((A Int)
         (B Int)
         (C Bool)
         (D Bool)
         (E Bool)
         (F Bool)
         (G Bool)
         (H Bool)
         (I Bool)
         (J Bool)
         (K Bool)
         (L Bool)
         (M Int)
         (N Int)
         (O Int)
         (P Int)
         (Q Bool)
         (R Bool)
         (S Bool)
         (T Bool)
         (U Int)
         (V Bool)
         (W Int)
         (X Int)
         (Y Bool)
         (Z Bool)
         (A1 Bool)
         (B1 Bool)
         (C1 Bool)
         (D1 Int)
         (E1 Int)
         (F1 Int))
  (let ((a!1 (= E1 (ite V (+ (- N 1) 1) 0)))
           (a!2 (= D1 (ite T (+ (- M 1) 1) 0)))
           (a!3 (= C1 (and (>= B 1) (>= A 1) (=> (>= E1 B) F) (=> (>= D1 A) E))))
           (a!4 (=> B1 (=> (>= E1 (+ B A)) E))))
       (=> (and (str_invariant X W V T S R Q P O N M L K I H G)
                (not (= A1 true))
                (not (= Z true))
                (not (= Y true))
                (not (= C true))
                a!1
                a!2
                a!3
                (= B1 (or C1 K))
                (= D a!4)
                (= B P)
                (= A O))
           (str_invariant F1 U J F E D C B A E1 D1 C1 B1 A1 Z Y)))
    ))


(check-sat)
(exit)
```
the output is : 
```
(error "
  failed to parse model from solver
")
```

I have updated hoice to this version : https://github.com/AdrienChampion/hoice/tree/warnings_and_rsmt2_update,
and the version of Z3 I used is 4.8.10.

**Comment by ryosu-sato** [comment]
This is due to the bug of Z3 ([#5146](https://github.com/Z3Prover/z3/issues/5146)),
and the bug is fixed by [cebf83c](https://github.com/Z3Prover/z3/commit/cebf83c4605a7d03fb6d90c803eee62232bb96e5).

**Comment by AdrienChampion** [comment]
This is a Z3-level problem, the CHCs above work on Z3 v4.8.7 which is the latest version that does not feature the recent bugs causing hoice to break.

Let me go ahead and close this, feel free to re-open/comment if necessary.
